### DISCUSSION TITLE
Update multiple rows without looping through

### DISCUSSION THREAD

**Discussion by anton-johansson** [prisma/prisma#19765] [Q&A]
This question has been asked before, see here: https://github.com/prisma/prisma/discussions/12389

But I think it's worth bringing it up again.

**EDIT**: In a hindsight, I should probably have just continued in the previous discussion. I as afraid that the discussion would remain far down in the list, but apparently, they are bumped up to the top whenever a message is added. Sorry!

## Current solution
If I want to update multiple rows in the same query, with different data, I have to loop and use `update` (or `upsert` if I also want to insert missing data), like this:

```typescript
const myId = 1;
const modifiedRows = [{
    myRowId: 1,
    myValue: "abc123",
}, {
    myRowId: 2,
    myValue: "def456",
}];

const promises = modifiedRows.map(({myRowId, myValue}) => transaction.myTable.update({
    data: {
        myValue,
    },
    where: {
        myId_myRowId: {
            myId,
            myRowId,
        },
    },
}));

await Promise.all(promises);
```

This is not optimal for performance. I assume there is some overhead with each query.

## An alternative solution (pure SQL)

I could easily write an SQL query that is more optimized and will perform a lot better. Something like this:

```sql
UPDATE  myTable
SET     myValue = x.myValue
FROM    MyTable myTable
INNER JOIN
(
        VALUES  (1, 'abc123')
        ,       (2, 'def456')
) x(myRowId, myValue)
        ON      x.myRowId = myTable.myRowId
WHERE   myTable.myId = 1
```

And I could easily build this up with TypeScript and call it with Prisma (**note:** SQL injection not considered):

```typescript
const myId = 1;
const modifiedRows = [{
    myRowId: 1,
    myValue: "abc123",
}, {
    myRowId: 2,
    myValue: "def456",
}];

const values = modifiedRows.map(({myRowId, myValue}, index) => `${index === 0 ? "VALUES" : ","} (${myRowId}, '${myValue}')`).join("\n");
const sql = `
UPDATE  myTable
SET     myValue = x.myValue
FROM    MyTable myTable
INNER JOIN
(
    ${values}
) x(myRowId, myValue)
        ON      x.myRowId = myTable.myRowId
WHERE   myTable.myId = ${myId}
`;

await transaction.$executeRawUnsafe(sql);
```

In my tests (that didn't have many rows at all, maybe around 10), I noticed a big difference in performance. Maybe it's because my test database is a bit slow, but that there is overhead here that can be reduced.

## Native Prisma solution

I'm not sure how, but it should be possible to create some kind of native Prisma query for this, right? I mean, the input should be relatively straight forward, and the generated SQL should be relatively forward as well (I'm surely missing something that causes complexity here, and if so, please let me know).

Maybe:
```typescript
// updateMany is already taken, but I can't think of any better name at the moment...
await transaction.myTable.updateMany({
    rows: [
        {
            data: {
                myValue: "abc123",
            },
            where: {
                myId_myRowId: {
                    myId: 1,
                    myRowId: 1,
                },
            },
        },
        {
            data: {
                myValue: "def456",
            },
            where: {
                myId_myRowId: {
                    myId: 1,
                    myRowId: 2,
                },
            },
        },
    ],
});
```

... or via the array:

```typescript
// updateMany is already taken, but I can't think of any better name at the moment...
await transaction.myTable.updateMany({
    rows: modifiedRows.map(({myRowId, myValue}) => ({
        data: {
            myValue,
        },
        where: {
            myId_myRowId: {
                myId,
                myRowId,
            },
        },
    }),
});
```

The only "catch" I see here is that it must somehow enforce that all items in the list use the same identifier keys, so that the `JOIN` in the SQL can be generated properly.

The same logic could be applied to `upsertMany` as well.

I would love to achieve higher performance, without sacrificing the amazing developer experience that Prisma provides. It would also handle SQL-injection for me out of the box, instead of using the unsafe `executeRawUnsafe` function.

What do you think? Am I missing something vital? Could this be added somehow?

**Comment by ludralph** [comment]
Hi @anton-johansson 

You are correct that using individual update or upsert queries in a loop can introduce overhead and impact performance when updating multiple rows with different data. Each query will have its own overhead in terms of network communication, query parsing, and database roundtrips. If you need to update multiple rows with different data and performance is a critical factor, using a custom SQL query like the one you shared can often result in better performance compared to executing individual Prisma queries. The SQL query you provided leverages a single UPDATE statement with a JOIN and a VALUES clause to efficiently update multiple rows in one operation. This approach reduces the overhead associated with multiple queries and can improve performance, especially when dealing with a large number of rows.

Your idea of having a native Prisma solution for efficiently updating multiple rows with different data is indeed valuable. While Prisma does not currently have a built-in method specifically designed for bulk updates of this nature, your proposed approach is a step in the right direction. Using a custom method like updateMany (as you suggested) or another suitable name, where you pass an array of objects containing the data and where clauses for each row, could be a viable solution. This would allow developers to update multiple rows in a single Prisma query, resulting in better performance compared to executing individual queries in a loop.

We do have a feature request to support updating multiple rows 

- https://github.com/prisma/prisma/issues/8238

You can add a 👍 to the feature request to help our engineers prioritize it while also documenting your use case in there.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by callmedukey** [comment]
Am using planetscale  and currently if I loop through about 20-30 items like

```js
await prisma.$transaction(async (tx) => {
    req.body.products.forEach(async (product) => {
      await tx.product.update({
        where: {
          id: product.id,
        },
        data: {
          someField: product.data,
        },
      });
    });
  });
```



It fails due to timeouts.... It works up till about 20 items, still stupidly slow and close to timeouts but anything more than 20 items just straight up fails

**Reply by callmedukey** [reply]
Works perfectly okay in a `new Promise.all(array.map())` and not a $transaction

**Reply by anton-johansson** [reply]
This might be a little bit misleading. It will go faster because it will exhaust more connections in the connection pool, which might not be what you want. Also, you might want/need a transaction to make sure everything is rolled back if any of the operations fail.

The original discussion point is related to making sure that we only get one **single** SQL query.

**Comment by aalin** [comment]
I just wrote a function to do this: https://gist.github.com/aalin/ea23b786e3d55329f6257c0f6576418b

And I use it like this:

```ts
prisma.$transaction([
  updateMulti("Category", ["parentId", "position"], categoryUpdates),
  updateMulti("Product", ["categoryId", "position"], productUpdates),
])
```

Where `categoryUpdates` and `productUpdates` are arrays of `[id, categoryId, position]`, like this:

```ts
const categoryUpdates: [number, number | null, number][] = [
  [6, null, 0],
  [11, 6, 0],
  [9, 11, 0],
  [10, 6, 1],
  [4, null, 1],
  [7, 4, 0],
  [8, 4, 1],
  [3, null, 2],
  [2, null, 3],
  [5, 2, 1],
  [1, null, 4]
]
```

This is printed to the logs:

```
prisma:query BEGIN
prisma:query UPDATE "Category" SET "parentId" = "t"."parentId", "position" = "t"."position" FROM (VALUES ($1,$2,$3),($4,$5,$6),($7,$8,$9),($10,$11,$12),($13,$14,$15),($16,$17,$18),($19,$20,$21),($22,$23,$24),($25,$26,$27),($28,$29,$30),($31,$32,$33)) AS t("id", "parentId", "position") WHERE "Category"."id" = "t"."id"
prisma:query UPDATE "Product" SET "categoryId" = "t"."categoryId", "position" = "t"."position" FROM (VALUES ($1,$2,$3),($4,$5,$6),($7,$8,$9),($10,$11,$12),($13,$14,$15),($16,$17,$18),($19,$20,$21),($22,$23,$24),($25,$26,$27),($28,$29,$30),($31,$32,$33),($34,$35,$36),($37,$38,$39),($40,$41,$42),($43,$44,$45),($46,$47,$48),($49,$50,$51),($52,$53,$54),($55,$56,$57),($58,$59,$60),($61,$62,$63),($64,$65,$66),($67,$68,$69),($70,$71,$72),($73,$74,$75),($76,$77,$78),($79,$80,$81),($82,$83,$84),($85,$86,$87),($88,$89,$90),($91,$92,$93),($94,$95,$96),($97,$98,$99),($100,$101,$102),($103,$104,$105),($106,$107,$108)) AS t("id", "categoryId", "position") WHERE "Product"."id" = "t"."id"
prisma:query COMMIT
```

It took around 300ms before when I did one update for each record, and now it takes about 10ms.

**Reply by anton-johansson** [reply]
Nice! I haven't tried it yet, but I'm guessing that the parameter value placeholder depends on which database/driver you are using. For MSSQL, I recall this being just `?`.

How about SQL injection? The `executeRawUnsafe` is subject to SQL injection, so you'd have to sanitize your values manually.

**Reply by aalin** [reply]
I use Postgres, but it might work with any other database with some modifications. About SQL injections, I'm not sure, but I think the difference between `executeRaw` and `executeRawUnsafe` is that the first one only accepts a tagged template string, while the the latter accepts a regular string. Since it's parameterized, the values should be escaped properly, according to this answer: https://github.com/prisma/prisma/discussions/20132#discussioncomment-6401410

**Reply by GGAlanSmithee** [reply]
dude @aalin a huge thanks for sharing this 🙇🏼

**Comment by HenryDotApp** [comment]
@anton-johansson How does your raw SQL solution compare to the [solution here](https://github.com/prisma/prisma/discussions/12389#discussioncomment-4856364) that uses `prisma.$transaction` on an array of the prisma updates?

**Reply by anton-johansson** [reply]
The solution mentioned in your link is the exact same problem that I'm mentioning in "Current solution" in my post, it does not matter if it's within a transaction or not. The problem is that each item in the array will generate it's own SQL query that needs to be passed to the database server, parsed by the database server, executed by the database server, and then responded to the client. This causes unwanted overhead. My raw SQL solution only generates one single SQL command, minimizing that overhead.

**Comment by peterHakio** [comment]
Good suggestion. May I suggest naming as **bulkUpdateMany** and **bulkUpsetMany**.

**Comment by carlosbaraza** [comment]
I'll just post my TypeScript solution to the problem using a helper called `bulkUpdate`. I tried to make it as generic as possible so it can serve a bunch of use cases (at least all the ones I commonly deal with):

```typescript
import { prisma } from './prisma-client';
import { PrismaPromise } from '@prisma/client';

export type BulkUpdateEntry = {
  id: number | string;
  [key: string]: number | string | boolean | Date;
};
export type BulkUpdateEntries = BulkUpdateEntry[];

export function bulkUpdate(
  tableName: string,
  entries: BulkUpdateEntries
): PrismaPromise<number> {
  if (entries.length === 0) return prisma.$executeRawUnsafe(`SELECT 1;`);

  const fields = Object.keys(entries[0]!).filter((key) => key !== 'id');
  const setSql = fields
    .map((field) => `"${field}" = data."${field}"`)
    .join(', ');

  const valuesSql = entries
    .map((entry) => {
      const values = fields.map((field) => {
        const value = entry[field];
        if (typeof value === 'string') {
          // Handle strings and escape single quotes
          return `'${value.replace(/'/g, "''")}'`;
        } else if (value instanceof Date) {
          // Convert Date to ISO 8601 string format
          return `'${value.toISOString()}'`;
        }
        // Numbers and booleans are used as-is
        return value;
      });

      return `('${entry.id}', ${values.join(', ')})`;
    })
    .join(', ');

  const sql = `
    UPDATE "${tableName}"
    SET ${setSql}
    FROM (VALUES ${valuesSql}) AS data(id, ${fields
    .map((field) => `"${field}"`)
    .join(', ')})
    WHERE "${tableName}".id::text = data.id;
  `;

  return prisma.$executeRawUnsafe(sql);
}
```

And the way to use it:
```typescript
    const batchedUsers = _.chunk(users, 500);
    const updates = batchedUsers.map((batchedUsers) => {
      const bulkUpdates: BulkUpdateEntries = batchedUsers.map((user) => {
        return {
          id: user.id,
          engagement: calculateEngagement(user),
        };
      });
      return bulkUpdate('user', bulkUpdates);
    });
    await prisma.$transaction(updates);
```


What used to take more than a minute now takes 5 seconds, which is good enough for my use case. There are some optimizations that could be done to the SQL query like avoiding the casting of the id to text. But I'll leave that to your own implementation.

**Reply by sebastianhader** [reply]
@carlosbaraza You could probably also add the type null as Value:
`else if (value === null) {
                        // Handle null values
                        return 'NULL'
                    }`

**Reply by carlosbaraza** [reply]
Good point, double equals would match undefined too. A version handling null/undefined:

```typescript
import { prisma } from './prisma-client';
import { PrismaPromise } from '@prisma/client';

export type BulkUpdateEntry = {
  id: number | string;
  [key: string]: number | string | boolean | Date;
};
export type BulkUpdateEntries = BulkUpdateEntry[];

export function bulkUpdate(
  tableName: string,
  entries: BulkUpdateEntries
): PrismaPromise<number> {
  if (entries.length === 0) return prisma.$executeRawUnsafe(`SELECT 1;`);

  const fields = Object.keys(entries[0]!).filter((key) => key !== 'id');
  const setSql = fields
    .map((field) => `"${field}" = data."${field}"`)
    .join(', ');

  const valuesSql = entries
    .map((entry) => {
      const values = fields.map((field) => {
        const value = entry[field];
        if (typeof value === 'string') {
          // Handle strings and escape single quotes
          return `'${value.replace(/'/g, "''")}'`;
        } else if (value instanceof Date) {
          // Convert Date to ISO 8601 string format
          return `'${value.toISOString()}'`;
        } else if (value == null) {
          // Handle null values or undefined
          return `NULL`;
        }
        // Numbers and booleans are used as-is
        return value;
      });

      return `('${entry.id}', ${values.join(', ')})`;
    })
    .join(', ');

  const sql = `
    UPDATE "${tableName}"
    SET ${setSql}
    FROM (VALUES ${valuesSql}) AS data(id, ${fields
    .map((field) => `"${field}"`)
    .join(', ')})
    WHERE "${tableName}".id::text = data.id;
  `;

  return prisma.$executeRawUnsafe(sql);
}
```

**Comment by wtkw** [comment]
@carlosbaraza I took your code and adapted it for Planetscale's MySQL: 

```typescript
import { PrismaPromise } from '@prisma/client'
import prisma from '../prisma'

export type BulkUpdateEntry = {
  [key: string]: number | string | boolean | Date | null
}
export type BulkUpdateEntries = BulkUpdateEntry[]

export function bulkUpdate(
  tableName: string,
  entries: BulkUpdateEntries,
  idCode: string,
  userId?: string,
): PrismaPromise<number> {
  if (entries.length === 0) return prisma.$executeRawUnsafe(`SELECT 1;`)

  const fields = Object.keys(entries[0]!).filter(
    (key) => key !== idCode && key !== 'userId',
  )

  const updateSetSql = fields
    .map((field) => `${field} = VALUES(${field})`)
    .join(', ')

  const userIdValueCond = userId ? `'${userId}', ` : ''

  const valuesSql = entries
    .map((entry) => {
      const values = fields.map((field) => {
        const value = entry[field]
        if (typeof value === 'string') {
          // Handle strings and escape single quotes
          return `'${value.replace(/'/g, "''")}'`
        } else if (value instanceof Date) {
          // Convert Date to ISO 8601 string format
          return `STR_TO_DATE('${value.toISOString()}', '%Y-%m-%dT%H:%i:%s.%fZ')`
        } else if (value == null) {
          return `NULL`
        }
        // Numbers and booleans are used as-is
        return value
      })

      return `(${userIdValueCond}'${entry[idCode]}', ${values.join(', ')})`
    })
    .join(', ')

  const sql = `INSERT INTO ${tableName} (${userId ? `userId,` : ``} ${idCode}, ${fields.map((field) => `${field}`).join(', ')})
  VALUES ${valuesSql}
  ON DUPLICATE KEY UPDATE ${updateSetSql};`

  return prisma.$executeRawUnsafe(sql)
}
```

You can specify if the `userId` exists in the table and in case of non-default primary id key `id` you can specify what it is.

Additionally, I added the `null` value support.

Usage example:

```typescript
const data = [
  {customId: 1, name: "Dave", date: new Date()},
  {customId: 2, name: "Steve", date: new Date()},
  {customId: 5, name: "John", date: new Date()}
]
```
```typescript
await bulkUpdate("UserTable", data, "customId", "cx1567128939234")
```

**Reply by kyle-belle** [reply]
i don't really understand how the `userId` parameter is used here? can you explain it a bit?

**Reply by kyle-belle** [reply]
### Update: I found out how to get the table names from generated prisma client
I made an update to this to make the typescript experience a little better

### Update: I added support for Json fields (but not sure how complete it is since I think strings values which contain double quotes need to be double escaped
```ts
import { prisma, Prisma } from "@app/db";
import { PrismaPromise } from "@prisma/client";

export function bulkUpdate<
  T extends Prisma.ModelName,
  IdField extends string = "id"
>(
  modelName: T,
  entries: (Partial<Prisma.TypeMap["model"][T]["payload"]["scalars"]> &
    Record<IdField, number | string>)[],
  idField: IdField = "id" as IdField,
  userId?: string
): PrismaPromise<number> {
  const tableName =
    Prisma.dmmf.datamodel.models.find((m) => m.name == modelName)?.dbName ||
    modelName;

  if (!tableName) {
    throw new Error(`Table name for model "${modelName} not found"`);
  }

  if (entries.length === 0) return prisma.$executeRawUnsafe(`SELECT 1;`);

  const fields = Object.keys(entries[0]!).filter(
    (key) => key !== idField && key !== "userId"
  ) as (keyof (typeof entries)[number])[];

  const updateSetSql = fields
    .map((field) => `${String(field)} = VALUES(${String(field)})`)
    .join(", ");

  const userIdValueCond = userId ? `'${userId}', ` : "";

  const valuesSql = entries
    .map((entry) => {
      const idValue = entry[idField];
      if (!idValue) {
        throw new Error(`All entries must have ${idField} property present.`);
      }

      const values = fields.map((field) => {
        const value = entry[field];

        if (typeof value === "string") {
          // Handle strings and escape single quotes
          return `'${value.replace(/'/g, "''")}'`;
        } else if (value instanceof Date) {
          // Convert Date to ISO 8601 string format
          return `STR_TO_DATE('${value.toISOString()}', '%Y-%m-%dT%H:%i:%s.%fZ')`;
        } else if (value == null) {
          return `NULL`;
        } else if (typeof value === "object") { // do this after checking null since `typeof null === "object"`
          return `'${JSON.stringify(value)}'`;
        }
        // Numbers and booleans are used as-is
        return value;
      });

      return `(${userIdValueCond}'${idValue}', ${values.join(", ")})`;
    })
    .join(", ");

  const sql = `INSERT INTO ${tableName} (${
    userId ? `userId,` : ``
  } ${idField}, ${fields.map((field) => `${String(field)}`).join(", ")})
  VALUES ${valuesSql}
  ON DUPLICATE KEY UPDATE ${updateSetSql};`;

  return prisma.$executeRawUnsafe(sql);
}
```
I still am not sure what the `userId` field does (i assume it's something very specific to your use case) but i still left it in my modified version.

And now with this i get helpful text completion and error messages from type script

Model Name Autocomplete
![Screenshot from 2024-07-08 18-11-11](https://github.com/prisma/prisma/assets/43228761/960479f5-7776-48d2-b8d3-669d875b5aff)
Model properties Autocomplete (scalar fields only)
![Screenshot from 2024-07-08 18-10-33](https://github.com/prisma/prisma/assets/43228761/94e9ceed-3c61-4048-9384-134a27500539)
Custom idField name (now required & throws typescript error if not present in entries)
![Screenshot from 2024-07-08 18-10-04](https://github.com/prisma/prisma/assets/43228761/048e62df-5dfb-4594-8805-02378e34c48e)

**Reply by carlosbaraza** [reply]
Yep, userId is something specific to me. It's unfortunately entangled cause I wrote that for a very specific purpose. But it's good you found a good generic prisma solution. I wish Prisma implemented it.

**Comment by peacefulotter** [comment]
Any update on this? I feel like prisma should include such function into their API since it's pretty common to update multiple rows at once...

**Comment by IRediTOTO** [comment]
no problem when use loop but I hope has a function like `prisma.transaction` to have beter perfomance

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Reply by peacefulotter** [reply]
This issue is still relevant and imho shouldn't be closed

**Reply by callmedukey** [reply]
Bump! This should not be closed. This is a common issue that does require attention. I get the Prisma team have priorities but I can't imagine why this isn't one of them considering how common it is to handle such a scenario.

**Reply by giovanioliani** [reply]
Yep... It's an absurdly common use case and my colleagues keep asking why I have to use an unsafe function that executes raw strings and I have to explain that the alternative is executing slow loops that insert (and then select, even if you're not using the result and with no option not to do so) records one by one.
